### DISCUSSION TITLE
8주차: 11장 API 리팩터링

### DISCUSSION THREAD

**Discussion by choisohyun** [frontend-book-study/refactoring-2#17] [리팩터링 2판]
인상깊었던 책의 내용과 인상깊었던 이유
현업 등 코드에 반영할 수 있는 방법과 생각
관련 아티클이나 이해를 돕는 자료를 첨부
새롭게 알게된 점이나 이해가 어려웠던 점

**Comment by choisohyun** [comment]
11.3 플래그 인수 제거하기

- 그동안 플래그 인수를 많이 사용했던 것 같은데... 플래그 인수가 있는 함수보다 함수를 두 개를 분리해서 사용하는 것이 의미가 명확할 수 있겠다고 느꼈습니다.

11.4 객체 통째로 넘기기

- 프론트에서 일반적으로 객체를 통째로 넘기는 함수의 경우 해당 도메인 함수 타입에 영향을 받는 경우가 많습니다. 그래서 객체를 통째로 넘기면 특정 도메인에 한정되는 느낌이라 pick을 해서 사용하곤 했습니다. 그런데 요게 오히려 로직 중복을 유발한다고 생각하니 또 그런 것 같네요. 그냥 통째로 쓰는 것도 나쁘지 않은 것 같습니다.

11.5 ~ 11.6 매개변수 <-> 질의 함수

- 보통 리팩토링할 대 질의 함수를 매개변수로 바꾸는 경우가 많았어서 함수 내에 다른 의존성이 들어있는 것을 좋아하지는 않았는데요, '쉽게'라는 관점에서 본다면 질의 함수로 두는 것도 (어느정도 한계점까지는) 괜찮다는 생각의 변화가 있었습니다.

11.9 함수를 명령으로 바꾸기

- command pattern을 처음 알게 되었습니다.

**Comment by LastCow9000** [comment]
- 일단 api의 책임에 관해 다시 한번 생각하게 되었습니다. api를 만들 때 한 가지의 역할만 하도록 command api와 query api를 잘 분리해서 작성해야겠습니다.
- 평소에 중복 제거를 위해 flag 인수를 써서 중복되는 함수를 합쳤었는데 이게 더 좋지 않은 결과(코드의 가독성)를 낸다는 것을 알게 되었습니다.
  - 다만 리터럴 값(true, false)이 아닌 의미를 나타내는 변수(isPremium)를 넘기면 그나마 괜찮다는 것을 알게되었습니다.
- 함수형 프로그래밍에서 중요하게 생각하는 참조투명성을 리팩터링 관점에서 다시 생각해볼 수 있었습니다.
  - 호출자의 책임 vs 함수의 책임

**Comment by hamo-o** [comment]
> 11.4 객체 통째로 넘기기
> 하나의 객체에서 값 두어 개를 가져와 인수로 넘기는 코드
> → 이 값들 대신 객체를 통째로 넘겨버리고 함수 본문에서 필요한 값들을 꺼내 쓰도록 하자.
> → 객체를 통째로 넘기면 변화에 대응하기 쉽기 때문이다.

자식 컴포넌트로 데이터를 넘겨줄 때 객체에서 값을 꺼내와 몇개만 프롭으로 넘기는 경우가 꽤 많았는데, 이 리팩터링을 적용해볼 수 있을 것 같습니다.

---

> 11.9 함수를 명령으로 바꾸기, 11.10 명령을 함수로 바꾸기
> 명령 객체(명령)란?
> → 함수를 그 함수만을 위한 객체 안으로 캡슐화하는 것
> → 즉 이 객체는 메서드 하나로 구성되며, 이 메서드를 실행하는 것이 객체의 목적이 된다.

여기서 갑자기 자바스크립트에서 함수를 일급으로 만들었기 때문에 일반적인 작업을 굳이 명령으로 리팩터링 할 필요가 없다는 부분이 이해하기 어려웠습니다. 그래서 우선 일급 객체와 함수에 대해 추가적으로 다시 알아봤습니다.

---

**자바스크립트의 일급 객체**

- 일급 객체란?
    - 객체를 값처럼 다룰 수 있다는 것
    - 변수에 담을 수 있다
    - 파라미터로 전달할 수 있다
    - 함수의 반환 값으로 사용될 수 있다
    - 동적으로 프로퍼티 생성이 가능하다
- 자바스크립트의 경우, 함수도 객체로 표현하기 때문에 **모든 함수는 일급 객체**이다.
- 자바스크립트가 일급 객체이기 때문에 아래와 같은 것들을 할 수 있다.
    - 콜백 함수
    - 고차함수
    - 클로저

그런데 이게 명령 패턴과 어떤 관련이 있을까요? 일단 명령 패턴이 무엇인지 몰라 이해가 어려웠는데, [여기](https://refactoring.guru/ko/design-patterns/command)서 잘 설명하고 있는 것 같아 스터디가 끝난 후 더 살펴보려 합니다.

간단히 알아보면, 아래와 같은 상황에서 클래스들이 어떠한 요청을 보내려고 했을 때 요 요청을 명령 객체로 따로 분리해낼 수 있을 것 같습니다.
<img width="579" alt="image" src="https://github.com/user-attachments/assets/d60d7709-734f-4ee4-ba0a-ff7e76a9a7bd">
<img width="570" alt="image" src="https://github.com/user-attachments/assets/3a1e03eb-e471-4a63-bdd4-7bdf2723a45e">

그런데 자바스크립트에는 함수 역시 일급 객체이므로 함수가 실행할 동작(명령)을 함수만으로 캡슐화할 수 있어 이런 이야기가 나온 것 같습니다. 아직 완전히 이해되지는 않아서 추가적으로 공부해볼 부분입니다!
